File: fixtures/coding/batch-ingestion-ts/tests/validate.test.ts
import { describe, expect, test } from "vitest";
import type { IngestRecord } from "../src/types";
import { validateRecord, validateBatch } from "../src/validate";

const record = (overrides: Partial<IngestRecord> = {}): IngestRecord => ({
  recordType: "click",
  ts: new Date(),
  payload: { page: "/home" },
  tags: new Set(["mobile"]),
  ...overrides,
});

describe("validateRecord", () => {
  test("valid record passes", () => {
    expect(validateRecord(record())).toEqual([]);
  });

  test("empty recordType fails", () => {
    const errors = validateRecord(record({ recordType: "" }));
    expect(errors.some((e) => e.field === "recordType")).toBe(true);
  });

  test("non-alphanumeric recordType fails", () => {
    const errors = validateRecord(record({ recordType: "click/event!" }));
    expect(errors.some((e) => e.field === "recordType")).toBe(true);
  });

  test("future timestamp fails", () => {
    const errors = validateRecord(record({ ts: new Date(Date.now() + 2 * 60 * 60 * 1000) }));
    expect(errors.some((e) => e.field === "ts")).toBe(true);
  });

  test("near-future timestamp allows one second skew", () => {
    const errors = validateRecord(record({ ts: new Date(Date.now() + 900) }));
    expect(errors.filter((e) => e.field === "ts")).toHaveLength(0);
  });

  test("empty payload fails", () => {
    const errors = validateRecord(record({ payload: {} }));
    expect(errors.some((e) => e.field === "payload")).toBe(true);
  });

  test("multiple errors on invalid record", () => {
    const errors = validateRecord(record({
      recordType: "",
      ts: new Date(Date.now() + 24 * 60 * 60 * 1000),
      payload: {},
    }));
    expect(errors.length).toBeGreaterThanOrEqual(3);
    const fields = new Set(errors.map((e) => e.field));
    expect(fields.has("recordType")).toBe(true);
    expect(fields.has("ts")).toBe(true);
    expect(fields.has("payload")).toBe(true);
  });

  test("validation does not mutate input", () => {
    const tags = new Set(["a"]);
    const r = record({ recordType: "test", payload: { key: "val" }, tags });
    const original = JSON.stringify({
      recordType: r.recordType,
      ts: r.ts.toISOString(),
      payload: r.payload,
      tags: [...r.tags],
    });
    validateRecord(r);
    expect(JSON.stringify({
      recordType: r.recordType,
      ts: r.ts.toISOString(),
      payload: r.payload,
      tags: [...r.tags],
    })).toBe(original);
  });
});

describe("validateBatch", () => {
  test("all valid", () => {
    const records = [record({ recordType: "a", payload: { x: 1 } }), record({ recordType: "b", payload: { y: 2 } })];
    const [valid, invalid] = validateBatch(records);
    expect(valid).toHaveLength(2);
    expect(invalid).toHaveLength(0);
  });

  test("some invalid", () => {
    const records = [
      record({ recordType: "valid", payload: { ok: true } }),
      record({ recordType: "", payload: { bad: "empty-type" } }),
      record({ recordType: "valid2", payload: { ok: false } }),
    ];
    const [valid, invalid] = validateBatch(records);
    expect(valid).toHaveLength(2);
    expect(invalid).toHaveLength(1);
    expect(invalid[0]?.[0].recordType).toBe("");
  });

  test("preserves valid-record order", () => {
    const records = [
      record({ recordType: "x", payload: { n: 1 } }),
      record({ recordType: "bad!", payload: { n: 2 } }),
      record({ recordType: "y", payload: { n: 3 } }),
    ];
    const [valid] = validateBatch(records);
    expect(valid.map((r) => r.payload.n)).toEqual([1, 3]);
  });

  test("does not mutate input array or records", () => {
    const records = [record({ recordType: "test", payload: { k: "v" } })];
    validateBatch(records);
    expect(records[0]?.recordType).toBe("test");
    expect(records[0]?.payload).toEqual({ k: "v" });
  });

  test("large input", () => {
    const records = Array.from({ length: 50 }, (_, i) => record({ recordType: `type${i}`, payload: { idx: i } }));
    const [valid, invalid] = validateBatch(records);
    expect(valid).toHaveLength(50);
    expect(invalid).toHaveLength(0);
  });
});

File: fixtures/coding/batch-ingestion-go/validate_test.go
package pipeline

import (
	"testing"
	"time"
)

func rec(overrides ...func(*Record)) Record {
	r := Record{
		RecordType: "click",
		TS:         time.Now().UTC(),
		Payload:    Payload{"page": "/home"},
		Tags:       StringSet{"mobile": {}},
	}
	for _, f := range overrides {
		f(&r)
	}
	return r
}

func hasField(errors []ValidationError, field string) bool {
	for _, err := range errors {
		if err.Field == field {
			return true
		}
	}
	return false
}

func TestValidateRecordValidPasses(t *testing.T) {
	if errors := ValidateRecord(rec()); len(errors) != 0 {
		t.Fatalf("expected no errors, got %#v", errors)
	}
}

func TestValidateRecordEmptyTypeFails(t *testing.T) {
	errors := ValidateRecord(rec(func(r *Record) { r.RecordType = "" }))
	if !hasField(errors, "record_type") {
		t.Fatalf("expected record_type error, got %#v", errors)
	}
}

func TestValidateRecordNonAlphanumericTypeFails(t *testing.T) {
	errors := ValidateRecord(rec(func(r *Record) { r.RecordType = "click/event!" }))
	if !hasField(errors, "record_type") {
		t.Fatalf("expected record_type error, got %#v", errors)
	}
}

func TestValidateRecordFutureTimestampFails(t *testing.T) {
	errors := ValidateRecord(rec(func(r *Record) { r.TS = time.Now().UTC().Add(2 * time.Hour) }))
	if !hasField(errors, "ts") {
		t.Fatalf("expected ts error, got %#v", errors)
	}
}

func TestValidateRecordNearFutureAllowsOneSecondSkew(t *testing.T) {
	errors := ValidateRecord(rec(func(r *Record) { r.TS = time.Now().UTC().Add(900 * time.Millisecond) }))
	if hasField(errors, "ts") {
		t.Fatalf("expected no ts error, got %#v", errors)
	}
}

func TestValidateRecordEmptyPayloadFails(t *testing.T) {
	errors := ValidateRecord(rec(func(r *Record) { r.Payload = Payload{} }))
	if !hasField(errors, "payload") {
		t.Fatalf("expected payload error, got %#v", errors)
	}
}

func TestValidateRecordMultipleErrors(t *testing.T) {
	errors := ValidateRecord(rec(func(r *Record) {
		r.RecordType = ""
		r.TS = time.Now().UTC().Add(24 * time.Hour)
		r.Payload = Payload{}
	}))
	if len(errors) < 3 || !hasField(errors, "record_type") || !hasField(errors, "ts") || !hasField(errors, "payload") {
		t.Fatalf("expected record_type, ts, payload errors, got %#v", errors)
	}
}

func TestValidateRecordDoesNotMutateInput(t *testing.T) {
	tags := StringSet{"a": {}}
	r := rec(func(r *Record) {
		r.RecordType = "test"
		r.Payload = Payload{"key": "val"}
		r.Tags = tags
	})
	ValidateRecord(r)
	if r.RecordType != "test" || r.Payload["key"] != "val" {
		t.Fatalf("record mutated: %#v", r)
	}
	if _, ok := r.Tags["a"]; !ok || len(r.Tags) != 1 {
		t.Fatalf("tags mutated: %#v", r.Tags)
	}
}

func TestValidateBatchAllValid(t *testing.T) {
	records := []Record{
		rec(func(r *Record) { r.RecordType = "a"; r.Payload = Payload{"x": 1} }),
		rec(func(r *Record) { r.RecordType = "b"; r.Payload = Payload{"y": 2} }),
	}
	valid, invalid := ValidateBatch(records)
	if len(valid) != 2 || len(invalid) != 0 {
		t.Fatalf("expected 2 valid/0 invalid, got %d/%d", len(valid), len(invalid))
	}
}

func TestValidateBatchSomeInvalid(t *testing.T) {
	records := []Record{
		rec(func(r *Record) { r.RecordType = "valid"; r.Payload = Payload{"ok": true} }),
		rec(func(r *Record) { r.RecordType = ""; r.Payload = Payload{"bad": "empty-type"} }),
		rec(func(r *Record) { r.RecordType = "valid2"; r.Payload = Payload{"ok": false} }),
	}
	valid, invalid := ValidateBatch(records)
	if len(valid) != 2 || len(invalid) != 1 || invalid[0].Record.RecordType != "" {
		t.Fatalf("unexpected split: valid=%#v invalid=%#v", valid, invalid)
	}
}

func TestValidateBatchPreservesOrder(t *testing.T) {
	records := []Record{
		rec(func(r *Record) { r.RecordType = "x"; r.Payload = Payload{"n": 1} }),
		rec(func(r *Record) { r.RecordType = "bad!"; r.Payload = Payload{"n": 2} }),
		rec(func(r *Record) { r.RecordType = "y"; r.Payload = Payload{"n": 3} }),
	}
	valid, _ := ValidateBatch(records)
	if len(valid) != 2 || valid[0].Payload["n"] != 1 || valid[1].Payload["n"] != 3 {
		t.Fatalf("valid order not preserved: %#v", valid)
	}
}

func TestValidateBatchDoesNotMutateInput(t *testing.T) {
	records := []Record{rec(func(r *Record) { r.RecordType = "test"; r.Payload = Payload{"k": "v"} })}
	ValidateBatch(records)
	if records[0].RecordType != "test" || records[0].Payload["k"] != "v" {
		t.Fatalf("input mutated: %#v", records)
	}
}

func TestValidateBatchLargeInput(t *testing.T) {
	records := make([]Record, 50)
	for i := range records {
		i := i
		records[i] = rec(func(r *Record) { r.RecordType = "type" + string(rune('A'+i%26)); r.Payload = Payload{"idx": i} })
	}
	valid, invalid := ValidateBatch(records)
	if len(valid) != 50 || len(invalid) != 0 {
		t.Fatalf("expected 50 valid/0 invalid, got %d/%d", len(valid), len(invalid))
	}
}
